lambda/matching.ml

```diff
@@ -320,9 +308,41 @@ end = struct
             { p with pat_desc = `Alias (Patterns.omega, id, str) }
             aliases rem
       | #view as view ->
-          let env = mk_alpha_env arg_id aliases vars in
-          ( (alpha env { p with pat_desc = view }, patl),
-            mk_action ~vars:(List.map snd env) )
+          (* We are doing two things here:
+             - we freshen the variables of the pattern, to
+               avoid reusing the same identifier in distinct exploded
+               branches
+             - we bind the variables in [alias] to the argument [arg]
+               (the other variables are bound in [view]); to avoid
+               code duplication if [arg] is itself not a variable, we
+               generate a binding for it, but only if the binding is
+               needed.
+
+             We are careful to avoid binding [arg] if not needed due
+             to the {!do_for_multiple_match} usage, which tries to
+             compile a tuple pattern [match e1, .. en with ...]
+             without allocating the tuple [(e1, .., en)].
+          *)
+          let env = List.map (fun id -> id, Ident.rename id) vars in
```

Comment:
Not all of the fresh new ids are being used, so you could delay the call to `Ident.rename` until you know that you're going to be using the id.
Also, I think the code would be easier to read without one more `env` variable with a different meaning from other `env`s in this file. (I know the name was there before your PR, but let's take the opportunity to change it!)

Reply:
I'm trying to apply your on-demand-rename suggestion:
- bad news: the resulting code is even more complex than the current one
- good news: this forced me to write a comment to explain what the function does in the first place, and I realized that we can probably simplify its interface a bit (it does not need to take the `patl` as part of the input as it ignores them, and in particular does not freshen any variables they would bind in the action, which is fine as the function is only called with omega-patterns there!).

Reply:
I pushed commits corresponding to those changes, let me know what you think.